File: internal/user/password_strength_test.go
package user

import "testing"

func TestIsCommonWeakPassword(t *testing.T) {
	cases := []struct {
		name     string
		password string
		want     bool
	}{
		{"exact match", "welcome1", true},
		{"case-insensitive match", "WELCOME1", true},
		{"mixed-case match", "Password1!", true},
		{"not on the list", "correct horse battery staple", false},
		{"password123 is deliberately not on the list", "password123", false},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			if got := isCommonWeakPassword(tc.password); got != tc.want {
				t.Errorf("isCommonWeakPassword(%q) = %v, want %v", tc.password, got, tc.want)
			}
		})
	}
}

func TestIsSingleRepeatedRune(t *testing.T) {
	cases := []struct {
		name     string
		password string
		want     bool
	}{
		{"all same letter", "aaaaaaaa", true},
		{"all same digit", "11111111", true},
		{"all spaces", "        ", true},
		{"single character", "a", true},
		{"empty string", "", true},
		{"mixed characters", "aaaaaaab", false},
		{"ordinary password", "correct horse battery staple", false},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			if got := isSingleRepeatedRune(tc.password); got != tc.want {
				t.Errorf("isSingleRepeatedRune(%q) = %v, want %v", tc.password, got, tc.want)
			}
		})
	}
}

func TestIsSequentialRun(t *testing.T) {
	cases := []struct {
		name     string
		password string
		want     bool
	}{
		{"ascending digits", "12345678", true},
		{"descending digits", "87654321", true},
		{"ascending letters", "abcdefgh", true},
		{"descending letters", "hgfedcba", true},
		{"single character", "a", false},
		{"empty string", "", false},
		{"broken run", "12345679", false},
		{"repeated, not sequential", "aaaaaaaa", false},
		{"ordinary password", "correct horse battery staple", false},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			if got := isSequentialRun(tc.password); got != tc.want {
				t.Errorf("isSequentialRun(%q) = %v, want %v", tc.password, got, tc.want)
			}
		})
	}
}

File: web/src/features/auth/passwordStrength.test.ts
import { describe, expect, it } from 'vitest'
import {
  isCommonWeakPassword,
  isSequentialRun,
  isSingleRepeatedRune,
  PASSWORD_REQUIREMENTS,
} from './passwordStrength'

describe('isCommonWeakPassword', () => {
  it('matches an exact entry', () => {
    expect(isCommonWeakPassword('welcome1')).toBe(true)
  })

  it('matches case-insensitively', () => {
    expect(isCommonWeakPassword('WELCOME1')).toBe(true)
    expect(isCommonWeakPassword('Password1!')).toBe(true)
  })

  it('does not match an uncommon password', () => {
    expect(isCommonWeakPassword('correct horse battery staple')).toBe(false)
  })

  it('deliberately does not list password123 — see the file-level comment', () => {
    expect(isCommonWeakPassword('password123')).toBe(false)
  })
})

describe('isSingleRepeatedRune', () => {
  it('is true for the same character repeated', () => {
    expect(isSingleRepeatedRune('aaaaaaaa')).toBe(true)
    expect(isSingleRepeatedRune('11111111')).toBe(true)
  })

  it('is false once a different character appears', () => {
    expect(isSingleRepeatedRune('aaaaaaab')).toBe(false)
  })

  it('is false for an ordinary password', () => {
    expect(isSingleRepeatedRune('correct horse battery staple')).toBe(false)
  })
})

describe('isSequentialRun', () => {
  it('is true for an ascending or descending run', () => {
    expect(isSequentialRun('12345678')).toBe(true)
    expect(isSequentialRun('87654321')).toBe(true)
    expect(isSequentialRun('abcdefgh')).toBe(true)
  })

  it('is false for a broken run', () => {
    expect(isSequentialRun('12345679')).toBe(false)
  })

  it('is false for a repeated (non-sequential) pattern', () => {
    expect(isSequentialRun('aaaaaaaa')).toBe(false)
  })

  it('is false for an ordinary password', () => {
    expect(isSequentialRun('correct horse battery staple')).toBe(false)
  })
})

describe('PASSWORD_REQUIREMENTS', () => {
  it('every requirement is met by a genuinely strong password', () => {
    for (const requirement of PASSWORD_REQUIREMENTS) {
      expect(requirement.met('correct horse battery staple')).toBe(true)
    }
  })

  it('the length requirement alone fails a short password', () => {
    const length = PASSWORD_REQUIREMENTS.find((r) => r.id === 'length')!
    expect(length.met('short1')).toBe(false)
  })

  it('the not-common requirement alone fails a blocklisted password', () => {
    const notCommon = PASSWORD_REQUIREMENTS.find((r) => r.id === 'not-common')!
    expect(notCommon.met('welcome1')).toBe(false)
  })

  it('the not-predictable requirement alone fails a sequential password', () => {
    const notPredictable = PASSWORD_REQUIREMENTS.find((r) => r.id === 'not-predictable')!
    expect(notPredictable.met('12345678')).toBe(false)
  })
})
